### ISSUE TITLE
Serving same gstream to multiple clients

### ISSUE THREAD

**Issue by EliDll** [janoglezcampos/GstreamerChromeBridge#2]
Hey!
Thank you for this nice work, I was able to display a Camera Feed in my Webapp  with the tcpBridgeAutoPipeline script.
Am I correct about the functionality that each time a Client issues a GET request it spawns a new gstreamer child process?
Is there any way to transmit the same gstreamer source to multiple clients at the same time with this?

**Comment by janoglezcampos** [comment]
Hi, you are right with your assumption, and with this code is not possible, the big deal is not that it creates one streamer pipeline each time you do a request, is that it creates a new tcp server each time, and you can't have more than one tcp server on the same port...

Im sorry but I can't code and try this days, I have college exams for 2 weeks, then I will be here to solve this, I think is a nice feature, btw, here are my first thoughts about how to solve this:

For the pipeline it should be easy to start it only once, you should move the code inside the callback of the tcpServer.listen, and run that piece of code outside, you can create a function like:

```
function startChildProcess(port){
var cmd = 'gst-launch-1.0';
        var args =
            ['videotestsrc',
                '!', 'video/x-raw,framerate=30/1,width=320,height=240',
                '!', 'videoconvert',
                '!', 'queue',
                '!', 'theoraenc',
                '!', 'queue',
                '!', 'm.', 'oggmux', 'name=m',
                '!', 'queue',
                '!', 'tcpclientsink', 'host=localhost',
                'port=' + port];

        var gstMuxer = child.spawn(cmd, args);

        gstMuxer.stderr.on('data', onSpawnError);
        gstMuxer.on('exit', onSpawnExit);
}
```

and then in the main level of the script run it. This works as a client, so it needs the tcp server to run before launching the pipeline. 
No need to kill the process manually, cause is a child process so it will be killed when the nodejs script is killed

Now comes the problem that im not really sure how to solve, and is starting only one tcpServer and sending the data of socket.on(data) to each http client.

TLDR:
New flow of the program:
1. Start 1 tcp server.
2. Start 1 pipeline that connects to that tcp server.
3. Start the http server, that will resend each tcp packet received by the server, to each client that made a request.

Ask me if you have any more questions, and I apologize now cause mby I take some time to answer this days.

**Comment by janoglezcampos** [comment]
It would look something like this, I didn't run it once, but mby you can have an idea of what I mean.

```
var express = require('express');
var http = require('http');
var net = require('net');
var child = require('child_process');

var expressApp = express();
var httpServer = http.createServer(expressApp);

var socket;

var tcpServer = net.createServer(function(sock){
    socket=sock;
});

tcpServer.maxConnections = 1;

var cmd = 'gst-launch-1.0';
var args =
    ['videotestsrc',
        '!', 'video/x-raw,framerate=30/1,width=320,height=240',
        '!', 'videoconvert',
        '!', 'queue',
        '!', 'theoraenc',
        '!', 'queue',
        '!', 'm.', 'oggmux', 'name=m',
        '!', 'queue',
        '!', 'tcpclientsink', 'host=localhost',
        'port=' + tcpServer.address().port];

var gstMuxer = child.spawn(cmd, args);

gstMuxer.stderr.on('data', onSpawnError);
gstMuxer.on('exit', onSpawnExit);

expressApp.get('/', function (req, res) {
    var date = new Date();

    res.writeHead(200, {
        'Date': date.toUTCString(),
        'Connection': 'close',
        'Cache-Control': 'private',
        'Content-Type': 'video/webm',
        'Server': 'CustomStreamer/0.0.1',
    });
    
    socket.on('data', function (data) {
        res.write(data);
    });
    
    socket.on('close', function (had_error) {
        res.end();
    });
});

httpServer.listen(8080);

function onSpawnError(data) {
    console.log(data.toString());
}

function onSpawnExit(code) {
    if (code != null) {
        console.log('GStreamer error, exit code ' + code);
    }
}

process.on('uncaughtException', function (err) {
    console.log(err);
});
```

**Comment by EliDll** [comment]
Thank you very much for the quick reply and nice description + code sample. I will try this out asap and get back to you. Best of luck with your exams!

**Comment by CzokNorris** [comment]
Hi, if someone figured this out, I would appreciate it a lot if you could post your code. 
The posted example does not work for me. I think tcpserver doesnt know the port at that time yet.
